libnmstate/route_rule.py

```diff
@@ -84,38 +102,64 @@ def is_valid(self, config_iface_routes):
-    def __init__(self, route_state, des_rule_state, cur_rule_state):
+    def __init__(self, route_state, des_rule_state, cur_rule_state, ifaces):
         self._config_changed = False
         self._cur_rules = defaultdict(set)
         self._rules = defaultdict(set)
         if cur_rule_state:
-            for rule_dict in _get_config(cur_rule_state):
-                rule = RouteRuleEntry(rule_dict)
-                self._cur_rules[rule.route_table].add(rule)
+            for entry in _get_config(cur_rule_state):
+                rl = RouteRuleEntry(entry)
+                self._cur_rules[rl.route_table].add(rl)
+                if not route_state or rl.is_valid(
+                    route_state.config_iface_routes
+                ):
+                    self._rules[rl.route_table].add(rl)
         if des_rule_state:
-            for rule_dict in _get_config(des_rule_state):
-                rule = RouteRuleEntry(rule_dict)
-                self._rules[rule.route_table].add(rule)
-            if self._rules != self._cur_rules:
-                self._config_changed = True
-        else:
-            # Discard invalid route rule when merging from current
-            for rules in self._cur_rules.values():
-                for rule in rules:
-                    if not route_state or rule.is_valid(
-                        route_state.config_iface_routes
-                    ):
-                        self._rules[rule.route_table].add(rule)
+            self._merge_rules(des_rule_state, route_state, ifaces)
 
     @property
     def _config(self):
         return _get_config(self._rules)
 
+    def _merge_rules(self, des_rule_state, route_state, ifaces):
+        """
+        Handle absent rules before adding desired rule entries to make sure
+        absent rule does not delete rule defined in desired state.
+        """
+        for entry in _get_config(des_rule_state):
+            rl = RouteRuleEntry(entry)
+            if rl.absent:
+                self._apply_absent_rules(rl, ifaces)
+        for entry in _get_config(des_rule_state):
+            rl = RouteRuleEntry(entry)
+            if not rl.absent:
+                self._rules[rl.route_table].add(rl)
+
+    def _apply_absent_rules(self, rl, ifaces):
+        """
+        Remove rules based on absent rules and treat missing property as
+        wildcard match.
+        """
+        absent_iface_table = rl.route_table
+        for route_table, rule_set in self._rules.items():
+            if absent_iface_table and absent_iface_table != route_table:
+                continue
+            new_rules = set()
+            for rule in rule_set:
+                if not rl.match(rule):
+                    new_rules.add(rule)
+            if new_rules != rule_set:
+                self._rules[route_table] = new_rules
+                for iface in ifaces.values():
+                    if iface.is_ipv4_enabled or iface.is_ipv6_enabled:
+                        iface.mark_as_changed()
```

Comment:
You are marking all interfaces as changed regardless they are holding route rule or not.

Since at this point, we don't know which interface will be used for saving this route rule, I think you can do
it in the `nm` level where you know current route rules and desire route rules or do it in `RouteRuleState.gen_metadata()`

Reply:
Done, it is being marked in `RouteRuleState.gen_metadata()`. Thanks and nice catch!